indexer/centers_table.cpp

```diff
@@ -0,0 +1,337 @@
+#include "indexer/centers_table.hpp"
+
+#include "indexer/coding_params.hpp"
+#include "indexer/geometry_coding.hpp"
+#include "indexer/point_to_int64.hpp"
+
+#include "coding/endianness.hpp"
+#include "coding/memory_region.hpp"
+#include "coding/reader.hpp"
+#include "coding/succinct_mapper.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
+#include "coding/writer.hpp"
+
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+
+#include "std/unordered_map.hpp"
+
+#include "3party/succinct/elias_fano.hpp"
+#include "3party/succinct/rs_bit_vector.hpp"
+
+namespace search
+{
+namespace
+{
+template <typename TCont>
+void EndiannesAwareMap(bool endiannesMismatch, CopiedMemoryRegion & region, TCont & cont)
+{
+  TCont c;
+  if (endiannesMismatch)
+  {
+    coding::ReverseMapVisitor visitor(region.MutableData());
+    c.map(visitor);
+  }
+  else
+  {
+    coding::MapVisitor visitor(region.ImmutableData());
+    c.map(visitor);
+  }
+
+  c.swap(cont);
+}
+
+// V0 of CentersTable.  Has the following format:
+//
+// File offset (bytes)  Field name          Field size (bytes)
+// 4                    positions offset    4
+// 8                    deltas offset       4
+// 12                   end of section      4
+// 16                   identifiers table   positions offset - 16
+// positions offset     positions table     deltas offset - positions offset
+// deltas offset        deltas blocks       end of section - deltas offset
+//
+// All offsets are in little-endian format.
+//
+// Identifiers table is a bit-vector with rank-select table, where set
+// bits denote that centers for the corresponding features are in
+// table.  Identifiers table is stored in a native endianness.
+//
+// Positions table is a elias-fano table, where each entry corresponds
+// to the start position of the centers block. Positions table is
+// stored in a native endianness.
+//
+// Deltas is a sequence of blocks, where each block is a sequence of
+// 64 varuints, where each varuint represents an encoded delta of the
+// center from some prediction. For the first center in the block map
+// base point is used as a prediction, for all other centers in the
+// block previous center is used as a prediction. This allows to
+// decode block of 64 consecutive centers at one syscall, and cache
+// them in RAM.
+class CentersTableV0 : public CentersTable
+{
+public:
+  static const uint32_t kFrequency = 64;
```

Comment:
I do not think it's an important point but constexpr could be used in more cases. If you want leave it as a constant leave it as is.